prov/shm/src/smr_progress.c

```diff
@@ -660,9 +644,76 @@ int smr_progress_unexp(struct smr_ep *ep, struct smr_ep_entry *entry)
 		ret = smr_progress_multi_recv(ep, &ep->trecv_queue, entry,
 					      total_len);
 		ret = ret ? ret : -FI_ENOMSG;
+	}
+
+	return ret;
+}
+
+int smr_progress_unexp_msg(struct smr_ep *ep, struct smr_ep_entry *entry)
+{
+	struct smr_match_attr match_attr;
+	struct smr_unexp_msg *unexp_msg;
+	struct dlist_entry *dlist_entry;
+	int ret = 0;
+
+	fastlock_acquire(&ep->region->lock);
+	if (ofi_cirque_isfull(ep->util_ep.rx_cq->cirq)) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+			"rx cq full\n");
+		ret = -FI_EAGAIN;
+		goto push_entry;
+	}
+
+	match_attr.addr = entry->addr;
+	dlist_entry = dlist_remove_first_match(&ep->unexp_msg_queue.list,
+					       ep->unexp_msg_queue.match_func,
+					       &match_attr);
+
+	if (!dlist_entry) {
+		ret = -FI_ENOMSG;
 		goto out;
 	}
 
+	unexp_msg = container_of(dlist_entry, struct smr_unexp_msg, entry);
+	ret = smr_process_matched_unexp(ep, unexp_msg, entry);
+
+push_entry:
+	freestack_push(ep->recv_fs, entry);
+out:
+	fastlock_release(&ep->region->lock);
+	return ret;
+}
+
+int smr_progress_unexp_tagged(struct smr_ep *ep, struct smr_ep_entry *entry)
```

Comment:
This function is basically identical to the one above.  Can we combine them and pass in the specific unexp_queue that we care about?

Reply:
That's a great idea! will update a new revision.